lib/extensions/descendant_loader.rb

```diff
@@ -0,0 +1,269 @@
+    end
+  end
+
+  # RubyParser is slow, so wrap it in a simple mtime-based cache.
+  module Cache
+    def cache_path
+      Rails.root.join('tmp/cache/sti_loader.yml')
+    end
+
+    def load_cache
+      return unless cache_path.exist?
+      data = YAML.load_file(cache_path)
+      return unless data && data.kind_of?(Hash) && data['@version'].to_i == CACHE_VERSION
+      data
+    end
+
+    def cache
+      @cache ||= load_cache || {'@version' => CACHE_VERSION}
+    end
+
+    def save_cache!
+      return unless @cache_dirty
+      cache_path.parent.mkpath
+      cache_path.open('w') do |f|
+        YAML.dump(cache, f)
+      end
+    end
+
+    def classes_in(filename)
+      t = File.mtime(filename)
+
+      if (entry = cache[filename])
+        return entry[:parsed] if entry[:mtime] == t
+      end
+
+      super.tap do |data|
+        @cache_dirty = true
+        cache[filename] = {:mtime => t, :parsed => data}
+      end
+    end
+  end
+
+  module Mapper
+    def class_inheritance_relationships
+      @class_inheritance_relationships ||= begin
+        children = Hash.new { |h, k| h[k] = [] }
+        Dir.glob(Rails.root.join("app/models/**/*.rb")) do |file|
+          classes_in(file).each do |define_scopes, search_scopes, name, sklass|
+            #binding.pry if name =~ /Vm\b/
+            possible_names = scoped_name(name, define_scopes)
+            possible_superklasses = scoped_name(sklass, search_scopes)
+
+            possible_superklasses.each do |possible_superklass|
+              children[possible_superklass].concat possible_names
+            end
+          end
+        end
+        children
+      end
+    end
+
+    def clear_class_inheritance_relationships
+      @class_inheritance_relationships = nil
+    end
+  end
+
+  include Parser
+  include Cache
+  include Mapper
+
+  def load_subclasses(parent)
+    names_to_load = class_inheritance_relationships[parent.to_s].dup
+    while (name = names_to_load.shift)
+      if (klass = name.safe_constantize) # this triggers the load
+        names_to_load.concat class_inheritance_relationships[name]
+      end
+    end
+  end
+
+  def scoped_name(name, scopes)
+    if name =~ /^::(.*)/
+      name = [$1]
+    else
+      scopes.map do |scope|
+        scope.empty? ? name : "#{scope}::#{name}"
+      end
+    end
+  end
+
+  module ArDescendantsWithLoader
+    def descendants
+      unless defined? @loaded_descendants
+        @loaded_descendants = true
+        DescendantLoader.instance.load_subclasses(self)
+      end
+
+      super
+    end
+  end
+
+  module AsDependenciesClearWithLoader
+    def clear
+      DescendantLoader.instance.clear_class_inheritance_relationships
+      super
+    end
+  end
+end
+
+ActiveRecord::Base.singleton_class.send(:prepend, DescendantLoader::ArDescendantsWithLoader)
+ActiveSupport::Dependencies.send(:prepend, DescendantLoader::AsDependenciesClearWithLoader)
+
+at_exit do
+  DescendantLoader.instance.save_cache!
+end
```

Comment:
Yes, we could/should probably arrange to pre-populate it in the build process.

It only gets written if it changed, and it's written as a non-append, so I *think* it should be okay for concurrent writes. (It would cause problems if it ended up written as an invalid yaml file... it going missing / becoming empty is obviously fine.)